### DISCUSSION TITLE
[Extension Question] Controlling ControlNet from AlwaysVisible script?

### DISCUSSION THREAD

**Discussion by bryan-d** [Mikubill/sd-webui-controlnet#1401] [Q&A] [answered]
Is it possible to update ControlNet parameters from the `process` method of an AlwaysVisible script? What I would like to do is be able to change the CNet input image based on keywords found in the prompt. Here is some example code which doesn't quite work:

```py
class my_extension(scripts.Script):
    def show(self, is_img2img):
        return scripts.AlwaysVisible

    def process(self, p: processing.StableDiffusionProcessing, *args):
        transformed_array, token = find_token(p.all_prompts)
        p.all_prompts = transformed_array

        if (token != None):
            layer = controlnet_module.get_all_units_in_processing(p)[0]
            if (layer.image == None) and (layer.enabled == True):
                image, mask = load_and_cache_image(os.path.realpath(os.path.join(__location__, f'../masks/{token}.png')))
                layer.image = {"image": image, "mask": mask}
                controlnet_module.update_cn_script_in_processing(p, [layer])
```

The code above has no issues aside from `controlnet_module.update_cn_script_in_processing(p, [layer])` not doing anything.

My guess is that you can only run the `update_cn_script_in_processing` method in the `run` function of a script. The issue is that in the `run` function, there doesn't seem to be `p.all_prompts`. There is `p.prompt`, but that can't be used for my situation because I need the final prompt after all replacements have been made by the dynamic-prompting extension. `p.all_prompts` returns exactly what I need, it's just that the `update_cn_script_in_processing` function doesn't seem to do anything when used in the `process` function.

Any workarounds?

**Comment by ljleb** [comment, marked as answer]
One way that is not ideal but would work, is to rename your script so that it comes lexicographically before controlnet. Then, you would need to import `controlnet_module` from within the `process` method so that controlnet can load before your script imports it.

Ideally you would call it before `process`, as after process it means the best thing we could do would be to call `process` twice, which could make the interaction with multiple extensions quite slow. Would it instead be possible for you to update the units whenever the user interacts with your gradio components?

**Reply by ljleb** [reply]
In fact, ideally the webui would allow alwayson scripts to specify which other extensions need to be ran before or after.

**Reply by bryan-d** [reply]
> One way that is not ideal but would work, is to rename your script so that it comes lexicographically before controlnet. Then, you would need to import `controlnet_module` from within the `process` method so that controlnet can load before your script imports it.
> 
> Ideally you would call it before `process`, as after process it means the best thing we could do would be to call `process` twice, which could make the interaction with multiple extensions quite slow. Would it instead be possible for you to update the units whenever the user interacts with your gradio components?

Works perfectly. This is for an internal project so not too worried about the non-ideal solution.

Thanks!
